src/QsCompiler/QirGeneration/Context.cs

```diff
@@ -1131,33 +1131,49 @@ Value BuildLoadForArg(QsArgumentTuple arg, Value value)
                 return args;
             }
 
-            void PopulateResultTuple(ResolvedType resultType, Value resultValue, Value resultTuplePointer, int item)
+            // result value contains the return value, and output tuple is the tuple where that value should be stored
+            void PopulateResultTuple(ResolvedType resultType, Value resultValue, Value outputTuple)
             {
                 var resultTupleType = this.LlvmStructTypeFromQsharpType(resultType);
+                Value GetOutputItem(int item) =>
+                    this.CurrentBuilder.GetElementPtr(
+                        resultTupleType,
+                        this.CurrentBuilder.BitCast(outputTuple, resultTupleType.CreatePointerType()),
+                        new[] { this.Context.CreateConstant(0L), this.Context.CreateConstant(item) });
+
                 if (resultType.Resolution is QsResolvedTypeKind.TupleType tupleType)
                 {
-                    // Here we'll step through and recurse
-                    var itemCount = tupleType.Item.Length;
                     // Start with 1 because the 0 element of the LLVM structures is the tuple header
-                    for (int i = 1; i <= itemCount; i++)
+                    for (int i = 1; i <= tupleType.Item.Length; i++)
                     {
-                        // TODO: complete implementation
+                        var resItem = this.CurrentBuilder.GetElementPtr(
+                             resultTupleType,
+                             resultValue,
+                             new[] { this.Context.CreateConstant(0L), this.Context.CreateConstant(i) });
+                        var itemOutputPointer = GetOutputItem(i);
+
+                        // if the subitem is a tuple we are good
+                        // if it is not then we need to get an output tuple to populate for the recursion to work... -> change recursion?
+                        if (tupleType.Item[i].Resolution.IsTupleType)
+                        {
+                            PopulateResultTuple(tupleType.Item[i], resItem, itemOutputPointer);
+                        }
+                        else
+                        {
+                            this.CurrentBuilder.Store(resItem, itemOutputPointer);
```

Comment:
You need to Load from resItem and store the result at itemOutputPointer; resItem is a pointer, not a final value.

Reply:
Thanks. How come LLVM/LLVM.NET doesn't complain here? There is a type distinction between the value and the pointer, and I have seen that in other places, the emission (and specifically Store) throws when there is a type mismatch. So how come it didn't throw here?

Reply:
Hmm... I don't know. Maybe I'm misreading and you don't need to Load, but I'm not sure.
Does the generated code look correct?

Reply:
This was me not thinking right about when this code would get invoked; the code only gets executed if a callable that has a tuple-valued output is passed as argument, and it indeed complains as it should.